cca_example_test.go

```diff
@@ -0,0 +1,187 @@
+}
+
+func (s symView) Dims() (r, c int) { return s.r, s.c }
+
+func (s symView) At(i, j int) float64 {
+	if i < 0 || s.r <= i {
+		panic("i out of bounds")
+	}
+	if j < 0 || s.c <= j {
+		panic("j out of bounds")
+	}
+	return s.sym.At(s.i+i, s.j+j)
+}
+
+func (s symView) T() mat64.Matrix { return mat64.Transpose{s} }
+
+func ExampleCanonicalCorrelations() {
+	// This example is directly analogous to Example 3.5 on page 87 of
+	// Koch, Inge. Analysis of multivariate and high-dimensional data.
+	// Vol. 32. Cambridge University Press, 2013. ISBN: 9780521887939
+
+	// bostonData is the Boston Housing Data of Harrison and Rubinfeld (1978)
+	n, _ := bostonData.Dims()
+	var xd, yd = 7, 4
+	// The variables (columns) of bostonData can be partitioned into two sets:
+	// those that deal with environmental/social variables (xdata), and those
+	// that contain information regarding the individual (ydata).
+	// Because the variables can be naturally partitioned in this way, these
+	// data are appropriate for canonical correlation analysis.
+	// The columns (variables) of xdata are, in order;
+	//  per capita crime rate by town,
+	//  proportion of non-retail business acres per town,
+	//  nitric oxide concentration (parts per 10 million),
+	//  weighted distances to Boston employment centres,
+	//  index of accessibility to radial highways,
+	//  pupil-teacher ratio by town, and
+	//  proportion of blacks by town.
+	xdata := bostonData.View(0, 0, n, xd)
+
+	// The columns (variables) of ydata are, in order;
+	//  average number of rooms per dwelling,
+	//  proportion of owner-occupied units built prior to 1940,
+	//  full-value property-tax rate per $10000, and
+	//  median value of owner-occupied homes in $1000s.
+	ydata := bostonData.View(0, xd, n, yd)
+
+	// For comparison, calculate the correlation matrix for the original data
+	var cor, cov mat64.SymDense
+	stat.CovarianceMatrix(&cov, bostonData, nil)
+	stat.CorrelationMatrix(&cor, bostonData, nil)
+
+	// Extract just those correlations that are between xdata and ydata.
+	var corRaw = symView{sym: &cor, i: 0, j: xd, r: xd, c: yd}
+
+	// Note that the strongest correlation between individual
+	// variables is 0.91 between the 5th variable of xdata (index of
+	// accessibility to radial highways) and the 3rd variable of ydata
+	// (full-value property-tax rate per $10000).
+	fmt.Printf("corRaw = %.4f", mat64.Formatted(corRaw, mat64.Prefix("         ")))
+
+	// Extract just those covariances that are between xdata and ydata,
+	// omitting those within xdata and within ydata.
+	var covRaw = symView{sym: &cov, i: 0, j: xd, r: xd, c: yd}
+
+	// Calculate the canonical correlations.
+	ccors, pVecs, qVecs, xsvd, ysvd, err := stat.CanonicalCorrelations(xdata, ydata, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Unpack xsvd and ysvd.
+	var xv, yv mat64.Dense
+	xv.VFromSVD(xsvd)
+	yv.VFromSVD(ysvd)
+	xs := xsvd.Values(nil)
+	ys := ysvd.Values(nil)
+	var xvs, yvs mat64.Dense
+	xvs.Clone(&xv)
```

Comment:
Because to calculate the canonical correlation matrix `corSph` below we need to multiply `xv` and `xvs` -- is there a better way to do this?